### PR TITLE
Fix gcc + binutils compilation.

### LINKED ISSUE #17294: Installation issue: bin utils
Hello, I am trying to install ```Spack``` on a cluster which seems to assume ```gcc@4.8.5``` as default. However I need Spack to be used by a code called ```PIConGPU``` whoose latest version makes use o ```gcc@7.4.0```.

I downloaded ``Spack``` through git, as git clone https://github.com/spack/spack.git volatile/src/spack, sourced the environment as source volatile/src/spack/share/spack/setup-env.sh and then I did ```spack compiler list | grep -q gcc@7.4.0 || spack install gcc@7.4.0 && spack load gcc@7.4.0 && spack compiler add``` once and another time with a completely fresh download of ```Spack``` I did ```spack install gcc@7.4.0~binutils```. The errors were identical as are shown below.

image

image

Please let mek now how could I find a work around for my problem.
Many thnaks,

Cristian

### LINKED ISSUE #17325: Installation issue: gcc@8.,3.0 with centos7 default compiler
### Steps to reproduce the issue

```console
$ spack install gcc@8.3.0
...
  >> 2426    checking for off_t... checking size of long long... configure: err
             or: in `spack-stage-gcc-8.3.0-6ypg
             gmswzvrzgw3lq6tovkv23mect6ee/spack-src/spack-build/gcc':
  >> 2427    configure: error: cannot compute sizeof (long long)
```

### Information on your system

$ spack debug report
* **Spack:** 0.15.0
* **Python:** 3.6.8
* **Platform:** linux-centos7-broadwell

### Additional information

Related to #3943 and #1866 which were not properly fixed but *avoided*.

PR #9024 solves this issue. (and I think +binutils should become the default).

Maintainer is @michaelkuhn .

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [ ] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by cedricchevalier19**
On Centos (RHEL) 7, a static libstdc++ is not available by default and
was required to compile stage1.

Now we use static libstdc++ only for stage > 1.

Almost like in PR #8972 binutils dependency has been modified to 'run', but also to 'build' instead of 'link'.

Fixes #17294
Fixes #17325

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -248,8 +248,13 @@ def configure_args(self):
             static_bootstrap_flags = '-static-libstdc++ -static-libgcc'
             binutils_options = [
                 '--with-sysroot=/',
-                '--with-stage1-ldflags={0} {1}'.format(
-                    self.rpath_args, static_bootstrap_flags),
+                # The following line causes linkage error on RHEL7
+                # static libstdc++ not being available by default.
+                # For stage1, we just want to produce a working c++ compiler,
+                # we can still rely on host infrastructure to do so.
+                # '--with-stage1-ldflags={0} {1}'.format(
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/gcc/package.py]: Hi @cedricchevalier19, could you remove the commented-out lines and update the comment:

> Don't set -static-libstdc++/-static-libgcc for --with-stage1-ldflags since a static libstdc++ is not available on all systems (RHEL7 for example). [THE REST IS COPIED] For stage1, we just want to produce a working c++ compiler, we can still rely on host infrastructure to do so.

Or however you want to rephrase it to make it clear without the commented-out code.

Thanks for the PR!

**Comment by scheibelp** [comment]
Hi @cedricchevalier19 do you think you'll have a chance to address https://github.com/spack/spack/pull/9024#discussion_r213015754 in the next week or so?

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -247,8 +247,11 @@ def configure_args(self):
             static_bootstrap_flags = '-static-libstdc++ -static-libgcc'
             binutils_options = [
                 '--with-sysroot=/',
-                '--with-stage1-ldflags={0} {1}'.format(
-                    self.rpath_args, static_bootstrap_flags),
+                # The following line causes linkage error on RHEL7
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/gcc/package.py]: @cedricchevalier19 You removed the line, but not this comment that is intended to refer to that line.

**Comment by cedricchevalier19** [comment]
GCC package.py has been updated but still does not work on Centos7.

**Comment by cedricchevalier19** [comment]
Fix  #17325

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 3]
```diff
@@ -115,7 +115,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('zstd', when='@10:')
     depends_on('iconv', when='platform=darwin')
     depends_on('gnat', when='languages=ada')
-    depends_on('binutils~libiberty', when='+binutils')
+    depends_on('binutils~libiberty', when='+binutils', type=('build', 'run'))
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py]:
```suggestion
    depends_on('binutils~libiberty', when='+binutils', type=('build', 'link', 'run'))
```
I would include `link` just to be sure.

**Comment by michaelkuhn** [comment]
@becker33 Looks like we broke CentOS/RHEL with #17280. Could you check whether aarch64 still works with these changes? I am currently doing some test builds on CentOS 6, CentOS 7 and Ubuntu 18.04 (all x86_64) and so far, everything looks good.

**Comment by becker33** [comment]
I will test on an aarch64 machine today and then approve/merge assuming it still works there.

**Comment by eugeneswalker** [comment]
This resolves the build problems I was having with gcc@7.3.0 on CentOS 7 using system compiler gcc@4.8.5.
